### PR TITLE
DDF-04413: Alerts can now dispatch queries to specific sources

### PR DESCRIPTION
**Pull request by leo-sakh**
#### What does this PR do?
Allows alerts to dispatch queries to only specific sources
#### Who is reviewing it? 
@bdeining 
@jrnorth 
@Corey-Collins 
@brianfelix 
#### Select relevant component teams: 
@codice/ui 
@codice/website 
#### Ask 2 committers to review/merge the PR and tag them here.

@millerw8

#### How should this be tested?
1. Schedule a query for specific source
2. Follow alert link to verify that only the desired source(s) has/have been queried
#### What are the relevant tickets?
For GH Issues:
Fixes: #4413 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by leo-sakh** [comment]
build now

**Comment by leo-sakh** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/component/alert/alert.js [hunk 1]
```diff
@@ -21,6 +21,16 @@ const router = require('../router/router.js')
 const user = require('../singletons/user-instance.js')
 const cql = require('../../js/cql.js')
 
+const sourceDataFromAlert = alert => {
+  const sources = alert.get('src')
+  const haveSources = sources && sources.length > 0
+
+  return {
+    federation: (haveSources && 'selected') || 'enterprise',
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/component/alert/alert.js]:
```suggestion
    federation: haveSources ? 'selected' : 'enterprise'
```
Probably just not what I'm used to, but it seems strange to write the logic this way versus using a ternary

**Comment by andrewkfiedler** [comment]
In order to keep this testable upstream and keep it in sync with downstream, should we update how we're doing alerts to take advantage of this?


Basically we would include the query details here:
https://github.com/codice/ddf/blob/master/ui/packages/catalog-ui-search/src/main/webapp/js/QueryPolling.js#L194

Maybe something like this:
````
wreqr.vent.trigger('alerts:add', {
       queryId: originalQuery.id,
       workspaceId: originalQuery.collection.parents[0].id,
       when: when,
       metacardIds: metacardIds,
       queryOptions:  {
                  federation: originalQuery.get('federation'),
                  src: originalQuery.get('src'),
        } 
})
```

The only gotcha with all of this is that the src name isn't guaranteed to be static over time.

**Comment by leo-sakh** [comment]
> In order to keep this testable upstream and keep it in sync with downstream, should we update how we're doing alerts to take advantage of this?
> 
> Basically we would include the query details here:
> https://github.com/codice/ddf/blob/master/ui/packages/catalog-ui-search/src/main/webapp/js/QueryPolling.js#L194
> 
> Maybe something like this:
> 
> ```
> wreqr.vent.trigger('alerts:add', {
>        queryId: originalQuery.id,
>        workspaceId: originalQuery.collection.parents[0].id,
>        when: when,
>        metacardIds: metacardIds,
>        queryOptions:  {
>                   federation: originalQuery.get('federation'),
>                   src: originalQuery.get('src'),
>         } 
> })
> ```
> The only gotcha with all of this is that the src name isn't guaranteed to be static over time.

@andrewkfiedler I have a couple concerns with this:

1. We're emitting events as strings. I'm the biggest offender when it comes to typos. We should probably create an event type, like an enum which we can reuse and serve as a single source of truce. This would also help us track down who's dispatching data and who's listening for it.
2. Data state, and it's mutations, are hard to visualize and reason about when they don't follow something like a flux pattern.

**Comment by djblue** [comment]
@leo-sakh I think @andrewkfiedler was saying we should update the existing code which uses the `wreqr.vent.trigger` call he linked. I agree with you that the wreqr event `'alerts:add'` should be part of an enumerated set of all known events, but I think that is a separate issue we should resolve in another PR.

**Comment by leo-sakh** [comment]
build now

**Comment by leo-sakh** [comment]
build now

**Comment by jrnorth** [comment]
I can hero this.

**Comment by jrnorth** [comment]
@leo-sakh I set up search notifications for a query against a single federated source. When I click on the alert, I see four requests to the `/cql` endpoint: one for the `cache` source, one for `ddf.distribution`, and one for each of my two federated sources.

**Comment by rzwiefel** [comment]
@leo-sakh Is this still needed?

**Comment by bennuttle** [comment]
I spoke to @leo-sakh and this change isn't needed anymore due to https://github.com/codice/ddf/pull/5589. This PR can safely be closed

**Pull request closed without merging**